src/crypto/rx-slow-hash.c

```diff
@@ -0,0 +1,356 @@
+} seedinfo;
+
+static CTHR_THREAD_RTYPE rx_seedthread(void *arg) {
+  seedinfo *si = arg;
+  randomx_init_dataset(rx_dataset, si->si_cache, si->si_start, si->si_count);
+  CTHR_THREAD_RETURN;
+}
+
+static void rx_initdata(randomx_cache *rs_cache, const int miners) {
+  if (miners > 1) {
+    unsigned long delta = randomx_dataset_item_count() / miners;
+    unsigned long start = 0;
+    int i;
+    seedinfo *si;
+    CTHR_THREAD_TYPE *st;
+    si = malloc(miners * sizeof(seedinfo));
+    if (si == NULL)
+      local_abort("Couldn't allocate RandomX mining threadinfo");
+    st = malloc(miners * sizeof(CTHR_THREAD_TYPE));
+    if (st == NULL) {
+      free(si);
+      local_abort("Couldn't allocate RandomX mining threadlist");
+    }
+    for (i=0; i<miners-1; i++) {
+      si[i].si_cache = rs_cache;
+      si[i].si_start = start;
+      si[i].si_count = delta;
+      start += delta;
+    }
+    si[i].si_cache = rs_cache;
+    si[i].si_start = start;
+    si[i].si_count = randomx_dataset_item_count() - start;
+    for (i=1; i<miners; i++) {
+      CTHR_THREAD_CREATE(st[i], rx_seedthread, &si[i]);
+    }
+    randomx_init_dataset(rx_dataset, rs_cache, 0, si[0].si_count);
+    for (i=1; i<miners; i++) {
+      CTHR_THREAD_JOIN(st[i]);
+    }
+    free(st);
+    free(si);
+  } else {
+    randomx_init_dataset(rx_dataset, rs_cache, 0, randomx_dataset_item_count());
+  }
+}
+
+static void rx_seedhash_int(rx_state *rx_sp, const uint64_t height, const char *hash, const int miners) {
+  randomx_flags flags = RANDOMX_FLAG_DEFAULT;
+  randomx_cache *cache;
+  CTHR_MUTEX_LOCK(rx_mutex);
+  cache = rx_sp->rs_cache;
+  if (rx_sp->rs_height != height || cache == NULL) {
+    if (use_rx_jit())
+      flags |= RANDOMX_FLAG_JIT;
+    if (cache == NULL) {
+      cache = randomx_alloc_cache(flags | RANDOMX_FLAG_LARGE_PAGES);
+      if (cache == NULL)
+        cache = randomx_alloc_cache(flags);
+      if (cache == NULL)
+        local_abort("Couldn't allocate RandomX cache");
+    }
+    randomx_init_cache(cache, hash, 32);
+    if (miners && rx_dataset != NULL)
+      rx_initdata(cache, miners);
+    rx_sp->rs_height = height;
+    if (rx_sp->rs_cache == NULL)
+      rx_sp->rs_cache = cache;
+  }
+  CTHR_MUTEX_UNLOCK(rx_mutex);
+  if (rx_vm != NULL)
+    randomx_vm_set_cache(rx_vm, rx_sp->rs_cache);
+}
+
+void rx_seedhash(const uint64_t height, const char *hash, const int miners) {
+  rx_state *rx_sp = &rx_s[rx_s_toggle];
+  rx_seedhash_int(rx_sp, height, hash, miners);
+}
+
+static char rx_althash[32];	// seedhash for alternate blocks
+
+void rx_alt_slowhash(const uint64_t mainheight, const uint64_t seedheight, const char *seedhash, const void *data, size_t length, char *hash) {
+  uint64_t s_height = rx_seedheight(mainheight);
+  int alt_toggle = (s_height & SEEDHASH_EPOCH_BLOCKS) == 0;
+  rx_state *rx_sp = &rx_s[alt_toggle];
+  if (rx_sp->rs_height != seedheight || rx_sp->rs_cache == NULL || memcmp(hash, rx_althash, sizeof(rx_althash))) {
```

Comment:
The `hash` in the `memcmp` here should be `seedhash`.

Reply:
Thanks, fixed